### PR TITLE
Hearing - Fix explosions not affecting hearing

### PR DESCRIPTION
**Pull request by johnb432**
**When merged this pull request will:**
- Title.

I have only done some testing with artillery shells and grenades, needs much more testing.

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: addons/hearing/functions/fnc_explosion.sqf [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * Example:
+ * [_projectile, [0, 0, 0], [0, 0, 0]] call ace_hearing_fnc_explosion
+ *
+ * Public: No
+ */
+
+params ["_projectile", "_pos"];
+
+// Don't allow for distances under 1
+private _distance = ((eyePos ACE_player) vectorDistance _pos) max 1;
```
Comment (PabstMirror) [inline addons/hearing/functions/fnc_explosion.sqf:23]: small optimization to ignore far away shots if we're adding this to all explosions, 
just guessing at what the cutoff should be

and block ear ringing if player in spectator (should match previous behavior)

```suggestion
private _distance = ((eyePos ACE_player) vectorDistance _pos) max 1;
if (_distance > 100) exitWith {}; // fast exit if explosion far away
if ((getNumber (configOf ACE_player>> "isPlayableLogic")) == 1) exitWith {};  
```

File: addons/hearing/XEH_postInit.sqf [hunk 2]
```diff
@@ -100,6 +91,12 @@ GVAR(lastPlayerVehicle) = objNull;
         call FUNC(updateHearingProtection);
     }, true] call CBA_fnc_addPlayerEventHandler;
 
+    addMissionEventHandler ["ProjectileCreated", {
+        params ["_projectile"];
+
+        _projectile addEventHandler ["Explode", {call FUNC(explosion)}];
```
Comment (PabstMirror) [inline addons/hearing/XEH_postInit.sqf:97]: strange locality issues with this (local 2 client MP)
server player's grenade and rockets don't trigger on the 2nd client
client's grenade does trigger on server but not the rocket

Server (q1):
```
Fired: [q1,"Throw","HandGrenadeMuzzle","HandGrenadeMuzzle","GrenadeHand","HandGrenade",164107: handgrenade_throw.p3d,q1]
Explode: [164107: handgrenade_throw.p3d,[1726.54,5510.9,5.53363],[0.0211423,-0.154011,0.0338565]]

Fired: [q1,"launch_RPG32_F","launch_RPG32_F","Single","R_PG32V_F","RPG32_F",164199: pg32v_rocket.p3d,q1]
Explode: [164199: pg32v_rocket.p3d,[1902.91,5710.63,19.3332],[134.031,108.958,4.30392]]

Fired: [q2,"launch_RPG32_F","launch_RPG32_F","Single","R_PG32V_F","RPG32_F",164203: pg32v_rocket.p3d,q2]

Fired: [q2,"Throw","HandGrenadeMuzzle","HandGrenadeMuzzle","GrenadeHand","HandGrenade",164205: handgrenade_throw.p3d,q2]
Explode: [164205: handgrenade_throw.p3d,[1718.96,5509.69,5.53045],[0.350581,-0.363032,-0.56609]]
```


Client (q2):
```
Fired: [q1,"Throw","HandGrenadeMuzzle","HandGrenadeMuzzle","GrenadeHand","HandGrenade",164212: handgrenade_throw.p3d,q1]

Fired: [q1,"launch_RPG32_F","launch_RPG32_F","Single","R_PG32V_F","RPG32_F",164389: pg32v_rocket.p3d,q1]

Fired: [q2,"launch_RPG32_F","launch_RPG32_F","Single","R_PG32V_F","RPG32_F",164450: pg32v_rocket.p3d,q2]
Explode: [164450: pg32v_rocket.p3d,[1998.34,5643.82,9.89063],[163.408,52.7211,-2.27858]]

Fired: [q2,"Throw","HandGrenadeMuzzle","HandGrenadeMuzzle","GrenadeHand","HandGrenade",164462: handgrenade_throw.p3d,q2]
Explode: [164462: handgrenade_throw.p3d,[1718.96,5509.69,5.53969],[0.34175,-0.295402,-0.014323]]
```

probably also effects https://github.com/acemod/ACE3/pull/9728

Comment (johnb432) [reply addons/hearing/XEH_postInit.sqf:97]: >server player's grenade and rockets don't trigger on the 2nd client

Server player's grenade does trigger on client for me. I'm running the latest profiling branch (2.16.151879) on both clients, which makes me wonder if some fixes related to throwables did something?

Regardless, I can reproduce the behaviour for rockets, where it only explodes on the client the rocket was fired from.
I think the fix is easy enough: We add the mission EH on each client, but only the client where the projectile is local adds the `Explode` EH and triggers a global event.

>probably also effects https://github.com/acemod/ACE3/pull/9728

Yea, definitely does. It means that we can't do things server-side only.

**Comment by johnb432** [comment]
This needs more work.

**Comment by johnb432** [comment]
I'm happy with it as it is. I've tested modded arty too and it works well enough all round.

File: addons/hearing/XEH_postInit.sqf [hunk 3]
```diff
@@ -100,6 +92,17 @@ GVAR(lastPlayerVehicle) = objNull;
...
+        params ["_projectile"];
+
+        if (!local _projectile) exitWith {};
+
+        // Rockets only explode on local clients
+        _projectile addEventHandler ["Explode", {
+            [QGVAR(explosion), _this] call CBA_fnc_globalEvent;
+        }];
+    }];
+
```
Comment (johnb432) [inline addons/hearing/XEH_postInit.sqf:105]: Can be moved to common, if the frag rewrite needs it.

**Comment by Dystopian** [comment]
I'm sorry I can't create correct Issue now. It looks like after this PR `Smoke Shell (White)` from `Mk6 Mortar` makes ears ringing.

**Pull request merged**